### ISSUE TITLE
Connecting to a BT device using console tools

### ISSUE THREAD

**Issue by vanfanel** [raspberrypi/linux#2633]
Hi there,

I have this Pi project where DBUS is not an option, but bluetooth connections are needed, in an embedded-like enviroment where services are mostly unwanted.

I was able to setup the BT hardware on the Pi by loading it's modules, and attaching it to hci:
```

sudo modprobe hci_uart
sudo modprobe btbcm
sudo hciattach /dev/ttyAMA0 bcm43xx 921600 noflow -
```
I can see it's working without problems with:
`sudo hciconfig -a`

I can easily bring up the device like this:
`sudo hciconfig hci0 up`

I can even scan devices with `sudo hcitool scan`, and I can see the target devices in range. Great so far!

The problem comes when I try to connect the Pi3b+ to them. There used to be `hidd --connect <address>`, but the hidd command has been deprecated in favor of some mumbo-jumbo using DBUS and systemd services. I guess a service is needed to manage connections, etc, but all I need here is to manually stablish an one-time BT connection manually. Simple.
So, what should I use in place of `hidd --connect <address>` now? I have tried `sudo hcitool cc <address>`, but it doesn't stablish any connections (connections can be verified with `hcitool con`).

**Comment by maxnet** [comment]
>I have this Pi project where DBUS is not an option

Don't think bluez works properly without dbus and bluetoothd daemon.
And hciattach/hcitool/hciconfig are also deprecated.

**Comment by vanfanel** [comment]
@maxnet : There's no way I will believe a service is necessary to stablish a connection. It's as if other network connections would need services to be stablished, that would make no sense. A service would take care of restoring connection settings on boot, saving them on shutdown, etc
We don't need a service running to stablish a wifi or ethernet connection, we have "ip" for that, and it's used by the networking services to do so.
So there must be a command to stablish BT connections manually, or things have just gone absurd.

Also, when I do

`sudo hciattach /dev/ttyAMA0 bcm43xx 921600 noflow -`

I get:

```
pi@raspberrypi:~ $ sudo hciattach /dev/ttyAMA0 bcm43xx 921600 noflow -
bcm43xx_init
Patch not found for BCM4345C0, continue anyway
Set Controller UART speed to 921600 bit/s
Device setup complete

```

Also, the Pi3b+ BT interface seems to be getting a WRONG adddress!

```
pi@raspberrypi:~ $ sudo hciconfig -a
hci0:	Type: Primary  Bus: UART
	BD Address: AA:AA:AA:AA:AA:AA  ACL MTU: 1021:8  SCO MTU: 64:1
	UP RUNNING 
	RX bytes:668 acl:0 sco:0 events:34 errors:0
	TX bytes:423 acl:0 sco:0 commands:34 errors:0
	Features: 0xbf 0xfe 0xcf 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH SNIFF 
	Link mode: SLAVE ACCEPT 
	Name: 'BCM4345C0'
	Class: 0x000000
	Service Classes: Unspecified
	Device Class: Miscellaneous, 
	HCI Version: 4.1 (0x7)  Revision: 0x0
	LMP Version: 4.1 (0x7)  Subversion: 0x6119
	Manufacturer: Broadcom Corporation (15)

```

Is that expected? Maybe that's why I can scan for BT devices but connection using hcitool does not work?
Do I need some kind of firmware? Shouldn't that be included in Raspbian then?

**EDIT: I have now installed the updated bluez-firmware package.** Now I get

```
pi@raspberrypi:~ $ sudo hciattach /dev/ttyAMA0 bcm43xx 921600 noflow -
bcm43xx_init
Flash firmware /lib/firmware/brcm/BCM4345C0.hcd

Set Controller UART speed to 921600 bit/s
Device setup complete
```

And I get:
```

pi@raspberrypi:~ $ sudo hciconfig -a
hci0:	Type: Primary  Bus: UART
	BD Address: 43:45:C0:00:1F:AC  ACL MTU: 1021:8  SCO MTU: 64:1
	UP RUNNING 
	RX bytes:708 acl:0 sco:0 events:39 errors:0
	TX bytes:522 acl:0 sco:0 commands:39 errors:0
	Features: 0xbf 0xfe 0xcf 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH SNIFF 
	Link mode: SLAVE ACCEPT 
	Name: 'BCM43455 37.4MHz Raspberry Pi 3+-0141'
	Class: 0x000000
	Service Classes: Unspecified
	Device Class: Miscellaneous, 
	HCI Version: 4.2 (0x8)  Revision: 0xfc
	LMP Version: 4.2 (0x8)  Subversion: 0x6119
	Manufacturer: Broadcom Corporation (15)

```

But even so, `sudo hcitool cc <address>` or `sudo hcitool cc <address>` still don't work, and they have to. 
```

pi@raspberrypi:~ $ sudo hcitool lecc E4:17:D8:15:0D:79
Could not create connection: Connection timed out

```
Any idea on what's wrong??

**Comment by maxnet** [comment]
1) The hciattach daemon that maintains the uart connection to the bluetooth controller is depreceated, and not necessary if you run your own embedded OS.
You can let the kernel take care of setting that up for you. However the downstream kernel currently lacks the right device-tree entries for that.
So that currently requires switching to upstream kernel, or some manual work.

However that will not solve your current problem, so will not go into the details.

2) If the bluetooth device you want to talk to requires any kind of pairing.
You need the bluetoothd daemon, and that will not run without dbus daemon.

It is not that dissimilair to other type of network connections.
Think wpa_supplicant for wifi...

**Comment by JamesH65** [comment]
@vanfanel Any furthur progress to report on this?

**Comment by vanfanel** [comment]
@JamesH65 : No. It seems I need a systemd dependant daemon running for BT pairing, so it won't work for me. 
Seems that systemd is winning the battle and GNU/Linux is depending on it for basic functions, that's a poisonous situation.
This can be closed, thanks.

**Comment by maxnet** [comment]
There is no dependency on systemd.
dbus != systemd
